### PR TITLE
feat!: expand environments specification

### PR DESCRIPTION
**Pull request by olevski**
This is a breaking change of the session launchers endpoints.

/deploy renku-ui=andrea/custom-launcher-jupyter-free

### CODE REVIEW

**Comment by olevski** [comment]
This replaces/supersedes #300.

**Review by Panaetius — CHANGES_REQUESTED** [review]

File: components/renku_data_services/session/api.spec.yaml [hunk 1]
```diff
@@ -437,13 +495,25 @@ components:
...
     EnvironmentKind:
       description: Kind of environment to use
       type: string
       enum:
-        - global_environment
-        - container_image
-      example: container_image
+        - GLOBAL
+        - CUSTOM
+      example: CUSTOM
```
[the remark this answers is not in the document]
Comment (olevski) [reply components/renku_data_services/session/api.spec.yaml:516]: They do know and that is why we cannot merge this. Andrea is working on adapting the ui to these changes.

File: components/renku_data_services/session/apispec.py [hunk 2]
```diff
@@ -72,12 +58,36 @@ class Environment(BaseAPISpec):
...
+    mount_directory: str = Field(
+        ...,
+        description="The location where the persistent storage for the session will be mounted, usually it should be identical to or a parent of the working directory",
+        min_length=1,
+    )
+    port: int = Field(
+        ...,
+        description="The TCP port (on any container in the session) where user requests will be routed to from the ingress",
+        gt=0,
+        le=65535,
```
[the remark this answers is not in the document]
Comment (olevski) [reply components/renku_data_services/session/apispec.py:85]: The sessions do not publish prometheus metrics. But yes we should blacklist some. I will add this.

File: components/renku_data_services/session/models.py [hunk 3]
```diff
@@ -2,61 +2,108 @@
...
+@dataclass(kw_only=True, frozen=True, eq=True)
+class UnsavedEnvironment(BaseEnvironment):
+    """Session environment model that has not been saved."""
+
+    port: int = 8888
+    description: str | None = None
+    working_directory: PurePosixPath = PurePosixPath("/home/jovyan/work")
+    mount_directory: PurePosixPath = PurePosixPath("/home/jovyan/work")
+    uid: int = 1000
+    gid: int = 1000
```
[the remark this answers is not in the document]
Comment (olevski) [reply components/renku_data_services/session/models.py:43]: I wanted to not have defaults on the saved environment. If I have defaults on the saved environment I could get in a case where I changed stuff in the DB and these fields were removed or renamed but the model is still applying these defaults even though the database is saying something else.

File: components/renku_data_services/session/db.py [hunk 4]
```diff
@@ -50,64 +56,105 @@ async def get_environment(self, environment_id: ULID) -> models.Environment:
...
+            created_by_id=user.id,
+            creation_date=datetime.now(UTC),
             description=new_environment.description,
             container_image=new_environment.container_image,
             default_url=new_environment.default_url,
-            created_by=models.Member(id=user.id),
-            creation_date=datetime.now(UTC).replace(microsecond=0),
+            port=new_environment.port,
+            working_directory=new_environment.working_directory.as_posix(),
+            mount_directory=new_environment.mount_directory.as_posix(),
```
[the remark this answers is not in the document]
Comment (olevski) [reply components/renku_data_services/session/db.py:78]: Hey Ralf that is a really nice way to do it. I will add this when I am back I will try to get the PR with the new amalthea operator in better shape first before I leave.

**Comment by olevski** [comment]
This change is part of the following stack:

- #360
    - #338 ◀

<sub>Change managed by [git-spice](https://abhinav.github.io/git-spice/).</sub>

File: components/renku_data_services/session/db.py [hunk 5]
```diff
@@ -299,29 +341,89 @@ async def update_launcher(
...
+                if launcher.environment.environment_kind == models.EnvironmentKind.GLOBAL:
+                    # A global environment is being replaced with a custom one
+                    if env_payload.get("environment_kind") == models.EnvironmentKind.GLOBAL:
+                        raise errors.ValidationError(
+                            message="When one global environment is being replaced with another in a "
+                            "launcher only the new global environment ID should be specfied",
+                            quiet=True,
+                        )
+                    env_payload["environment_kind"] = models.EnvironmentKind.CUSTOM.value
+                    env_payload_valid = apispec.EnvironmentPostInLauncher.model_validate(env_payload)
```
Comment (Panaetius) [inline components/renku_data_services/session/db.py:390]: thought: to be honest this sucks a bit, it is a lot of business type logic in the database repo. But in my PR that I based on top of yours, I tried to come up with a cleaner solution (where the logic on if we need to create a new environment, switch from global to global or not, etc. is handled in a core.py and we don't need to depend on the apispec outside of blueprints) but I could not come up with a solution that's any cleaner. 
The only thing that came to mind is not allowing posting an environment body here at all, only an ID, and requiring clients to creates environments by posting to an environment endpoint first before patching the launcher. But that's not as nice from the UX perspective of clients.

This method here is rather long though and I wonder if it should be split up to deal with e.g. environments and resource classes separately.

Comment (olevski) [reply components/renku_data_services/session/db.py]: Now it is cleaned up. Check the last commit in this PR.

**Review by Panaetius — APPROVED** [review]

**Review by Panaetius — APPROVED** [review]

**Pull request merged**